notes/fomega/evaluation/builtin-evaluation.md

```diff
@@ -0,0 +1,347 @@
+performs a computation, and returns an integer; `b₂` takes one
+integer, performs some computation, and returns a function which can
+be applied to another integer.  The partial applications `[(builtin
+b₁) (con 1)]` and `[(builtin b₂) (con 1)]` appear identical, but the
+first does not perform any computation whereas the second does
+(and may fail when it does so).
+
+
+We've recently had a discussion about whether we want to allow type
+and term arguments to be interleaved, allowing us to have builtins
+with non-canonical types (and possibly higher-rank types, although I'm
+still confused about this).  The internal machinery doesn't really
+care about this, but it does present some problems.
+
+One possible issue is that if we allow type arguments between term
+arguments there's the question of how we erase them.  If we allow such
+things, then we can have a polymorphic partially-applied builtin, and
+such a thing could be passed to a function which could also accept
+type-abstracted lambdas, and such a function would have to deal
+uniformly with both possibilities.  The difficulty is that we no longer
+have the value restriction, so we have to erase abstractions to
+`delay` and apply `force` when type instantiation occurs.  Since
+there's no possibility of computation occurring when a
+partially-applied builtin is instantiated, it seems strange to
+unnecessarily wrap them in a `delay`. [But what happens if the final
+argument is a type argument, ie, if a builtin returns a polymorphic
+value?  We currently require all builtin arguments to be values and
+builtins are supposed to be parametrically polymorphic _and_ return
+values, so can a builtin really return something that can compute when
+instantiated?  Perhaps it can: maybe we can pass `Λt:*.error` (which
+erases to `delay error`) to the identity function?]
+
+Anyway, perhaps we could modify `force` so that it'll act on _any_
+value, doing nothing when it's not wrapped in `delay`; we would then
+be able to completely erase type arguments to builtins except perhaps
+ones at the very end.  (I'm now wondering what would happen if you
+force a _term_: could that ever happen?)
+
+Roman has a number of other proposals, including erasing type arguments
+of builtins to `()` and prohibiting final quantifiers altogether.
+
+### Do we want non-prenex signatures?
+
+It's maybe worth asking whether we need such things.  Is there a real
+use case where they'd be useful?  Do they allow us to compile even
+more Haskell into Plutus Core?  It appears that there's not a huge
+amount of extra implementation effort involved though, so they
+wouldn't be too difficult to support.
+
+One thing that worries me a little is how signatures for builtins with
+interleaved type and term arguments  affect the specification.  It
+might be quite tricky just to write down a generic instance of such a
+signature, for example.
+
+What about formalisation? Would this complicate things in the Agda
```

Comment:
This will require some work definitely. It took me a few iterations to have a practical treatment of saturated non-interleaved builtins.